tests/core/fused_gromov_wasserstein_test.py

```diff
@@ -0,0 +1,249 @@
+                        rtol=1e-02, atol=1e-02)
+
+  def test_adaptive_threshold_fused(self):
+    """Checking solution is improved with smaller threshold for convergence."""
+    geom_x = pointcloud.PointCloud(self.x, self.x)
+    geom_y = pointcloud.PointCloud(self.y, self.y)
+    geom_xy = pointcloud.PointCloud(self.x_2, self.y_2)
+    # without warm start for calls to sinkhorn
+    def loss_thre(threshold):
+      return gromov_wasserstein.gromov_wasserstein(
+          geom_xx=geom_x, geom_yy=geom_y, geom_xy=geom_xy, fused_penalty=self.fused_penalty_2, a=self.a, b=self.b,
+          epsilon=.1, threshold=threshold).reg_gw_cost
+    self.assertGreater(loss_thre(.1), loss_thre(.001))
+    self.assertGreater(loss_thre(.001), loss_thre(.00001))
+
+  @parameterized.parameters([True], [False])
+  def test_gradient_fused_gromov_wasserstein_penalty(self, lse_mode):
+    """Test gradient w.r.t. penalty."""
+
+    def reg_gw(cx, cy, cxy, fused_penalty, a, b, implicit):
+      geom_x = geometry.Geometry(cost_matrix=cx)
+      geom_y = geometry.Geometry(cost_matrix=cy)
+      geom_xy = geometry.Geometry(cost_matrix=cxy)
+      sinkhorn_kwargs = {'implicit_differentiation': implicit,
+                         'max_iterations': 1001, 'lse_mode': lse_mode}
+      return gromov_wasserstein.gromov_wasserstein(
+        geom_x, geom_y, geom_xy=geom_xy, fused_penalty=fused_penalty, a=a, b=b, epsilon=1.0, max_iterations=10,
+        sinkhorn_kwargs=sinkhorn_kwargs).reg_gw_cost
+
+    grad_matrices = [None, None]
+    for i, implicit in enumerate([True, False]):
+      reg_gw_and_grad = jax.value_and_grad(reg_gw, argnums=(3,))
+      _, grad_reg_gw = reg_gw_and_grad(
+        self.cx, self.cy, self.cxy, self.fused_penalty, self.a, self.b, implicit)
+      grad_matrices[i] = grad_reg_gw
+      self.assertIsNot(jnp.any(jnp.isnan(grad_reg_gw[0])), True)
+    self.assertAllClose(grad_matrices[0][0], grad_matrices[1][0],
+                        rtol=1e-02, atol=1e-02)
+
+  def test_effect_fused_penalty(self):
+
+    def reg_fgw(x, y, x_2, y_2, fused_penalty, a, b):
+      geom_x = pointcloud.PointCloud(x)
+      geom_y = pointcloud.PointCloud(y)
+      geom_xy = pointcloud.PointCloud(x_2, y_2)
+      sinkhorn_kwargs = {'max_iterations': 1001}
+      return gromov_wasserstein.gromov_wasserstein(
+        geom_x, geom_y, geom_xy=geom_xy, fused_penalty=fused_penalty, a=a, b=b, epsilon=1.0,
+        sinkhorn_kwargs=sinkhorn_kwargs)
+
+    def reg_gw(x, y, a, b):
+      geom_x = pointcloud.PointCloud(x)
+      geom_y = pointcloud.PointCloud(y)
+      sinkhorn_kwargs = {'max_iterations': 1001}
+      return gromov_wasserstein.gromov_wasserstein(
+        geom_x, geom_y, a=a, b=b, epsilon=1.0,
+        sinkhorn_kwargs=sinkhorn_kwargs)
+
+    fgw_output = reg_fgw(self.x, self.y, self.x_2, self.y_2, self.fused_penalty, self.a, self.b)
+    gw_output = reg_gw(self.x, self.y, self.a, self.b)
+    self.assertNotAlmostEqual(fgw_output.reg_gw_cost, gw_output.reg_gw_cost)
```

Comment:
my guess is that you could test that fgw_output.reg_gw_cost should be bigger than gw_output.reg_gw_cost

Reply:
True, thank you, added that to the follow-up pull request.